tests/test_variant_algebra.py

```diff
@@ -314,3 +314,260 @@ def test_pin_run_as_build():
+
+    print(res)
+    print(res2)
+
+    assert res2["python"] == migration_py39["__migrator"]["ordering"]["python"]
+    # assert that we've ordered the numpy bits properly
+    assert res2["numpy"] == [
+        "1.16",
+        "1.100",
+        "1.16",
+        "1.16",
+        "1.18",
+    ]
+
+    res3 = variant_add(base, migration_py39)
+    print(res3)
+    assert res3["python"] == [
+        "3.6.* *_cpython",
+        "3.9.* *_cpython",  # newly added
+        "3.7.* *_cpython",
+        "3.8.* *_cpython",
+    ]
+    # The base doesn't have an entry for numpy
+    assert "numpy" not in res3
+
+
+def test_variant_key_remove():
+    base = parse_variant(
+        dedent(
+            """
+    python:
+      - 3.6.* *_cpython
+      - 3.8.* *_cpython
+      - 3.6.* *_73_pypy
+    numpy:
+      - 1.16
+      - 1.16
+      - 1.18
+    python_impl:
+      - cpython
+      - cpython
+      - pypy
+    zip_keys:
+      -
+        - python
+        - numpy
+        - python_impl
+    """
+        )
+    )
+    removal = parse_variant(
+        dedent(
+            """
+            __migrator:
+                operation: key_remove
+                primary_key: python
+                ordering:
+                    python:
+                        - 3.6.* *_73_pypy
+                        - 3.6.* *_cpython
+                        - 3.7.* *_cpython
+                        - 3.8.* *_cpython
+                        - 3.9.* *_cpython
+            python:
+              - 3.6.* *_cpython
+            """
+        )
+    )
+
+    res = variant_add(base, removal)
+
+    assert res["python"] == ["3.6.* *_73_pypy", "3.8.* *_cpython"]
+    assert res["numpy"] == ["1.18", "1.16"]
+    assert res["python_impl"] == ["pypy", "cpython"]
+
+
+@pytest.mark.parametrize(
+    "platform,arch", [["osx", "64"], ["osx", "arm64"], ["linux", "64"]]
+)
+def test_variant_remove_add(platform, arch):
+    from conda_build.config import Config
+
+    config = Config(platform=platform, arch=arch)
+    base = parse_variant(
+        dedent(
+            """
+            python:
+              - 3.7.* *_cpython   # [not (osx and arm64)]
+              - 3.8.* *_cpython
+              - 3.6.* *_73_pypy   # [not (win64 or (osx and arm64))]
+
+            numpy:
+              - 1.16       # [not (osx and arm64)]
+              - 1.16
+              - 1.18       # [not (win64 or (osx and arm64))]
+
+            python_impl:
+              - cpython    # [not (osx and arm64)]
+              - cpython
+              - pypy       # [not (win64 or (osx and arm64))]
+
+
+            zip_keys:
+              -
+                - python
+                - numpy
+                - python_impl
+            """
+        ),
+        config=config,
+    )
+
+    remove = parse_variant(
+        dedent(
+            """
+            __migrator:
+                operation: key_remove
+                primary_key: python
+            python:
+              - 3.8.* *_cpython
```

Comment:
Can this just be a removal for `osx and arm64` only? If so, we can reduce the 3 operations to one removal and one addition

Reply:
i think that can work, mostly keeping it with a full remove, add, add so that the contents in the migrators more closely resembles what we'd eventually have in conda_build_config.yaml

Reply:
True. Also, we can keep `3.8` for arm64 for a while if it makes it easier for you.

Reply:
Added a test to show that works fine too